nuxeo-features/nuxeo-platform-comment/nuxeo-platform-comment/src/main/java/org/nuxeo/ecm/platform/comment/impl/TreeCommentManager.java

```diff
@@ -0,0 +1,479 @@
+     */
+    protected void checkUpdateCommentPermissions(CoreSession session, Comment comment) {
+        NuxeoPrincipal principal = session.getPrincipal();
+        if (!principal.isAdministrator() && !comment.getAuthor().equals(principal.getName())) {
+            throw new CommentSecurityException(String.format("The user %s cannot edit comments of document %s",
+                    principal.getName(), comment.getParentId()));
+        }
+    }
+
+    /**
+     * @return the external document model for the given {@code entityId}, if it exists, otherwise throws a
+     *         {@link CommentNotFoundException}
+     */
+    @SuppressWarnings("unchecked")
+    protected DocumentModel getExternalCommentModel(CoreSession session, String entityId) {
+        PageProviderService ppService = Framework.getService(PageProviderService.class);
+        Map<String, Serializable> props = singletonMap(CORE_SESSION_PROPERTY, (Serializable) session);
+        PageProvider<DocumentModel> pageProvider = (PageProvider<DocumentModel>) ppService.getPageProvider(
+                GET_COMMENT_PAGE_PROVIDER_NAME, Collections.emptyList(), 1L, 0L, props, entityId);
+        List<DocumentModel> documents = pageProvider.getCurrentPage();
+        if (documents.isEmpty()) {
+            throw new CommentNotFoundException(String.format("The external comment %s does not exist.", entityId));
+        }
+        return documents.get(0);
+    }
+
+    /**
+     * Remove the comment of the given {@code documentRef}
+     *
+     * @param s the core session
+     * @param documentRef the documentRef of the comment document model to delete
+     */
+    protected void removeComment(CoreSession s, DocumentRef documentRef) {
+        NuxeoPrincipal principal = s.getPrincipal();
+        CoreInstance.doPrivileged(s, session -> {
+            DocumentRef ancestorRef = getRootDocument(session, documentRef);
+            DocumentModel commentDocModel = session.getDocument(documentRef);
+            Serializable author = commentDocModel.getPropertyValue(COMMENT_AUTHOR);
+            if (!(principal.isAdministrator() //
+                    || author.equals(principal.getName()) //
+                    || session.hasPermission(principal, ancestorRef, EVERYTHING))) {
+                throw new CommentSecurityException(String.format("The user %s cannot delete comment of the document %s",
+                        principal.getName(), ancestorRef));
+            }
+            DocumentModel parent = session.getDocument(commentDocModel.getParentRef());
+            commentDocModel.detach(true);
+            session.removeDocument(documentRef);
+            notifyEvent(session, CommentEvents.COMMENT_REMOVED, parent, commentDocModel);
+        });
+
+    }
+
+    /**
+     * @return the root document of the comments hierarchy for the given {@code documentRef}
+     */
+    protected DocumentRef getRootDocument(CoreSession s, DocumentRef documentRef) {
```

Comment:
Thx, you are right, after reflection this method will removed because it's can easy be replaced by the  `getTopLevelComment` onces, it will save some code.